File: lib/vxtp/rifs.c
/*
Copyright (c) 2019-2022 Andreas T Jonsson

This software is provided 'as-is', without any express or implied
warranty. In no event will the authors be held liable for any damages
arising from the use of this software.

Permission is granted to anyone to use this software for any purpose,
including commercial applications, and to alter it and redistribute it
freely, subject to the following restrictions:

1. The origin of this software must not be misrepresented; you must not
   claim that you wrote the original software. If you use this software
   in a product, an acknowledgment in the product documentation would be
   appreciated but is not required.
2. Altered source versions must be plainly marked as such, and must not be
   misrepresented as being the original software.
3. This notice may not be removed or altered from any source distribution.
*/

#include "vxtp.h"
#include "crc32.h"

#include <assert.h>
#include <string.h>
#include <stdio.h>

#define BUFFER_SIZE 1024

VXT_PACK(struct rifs_packet {
    vxt_byte packetID[2];   // 'K', 'Y': Request from server.
                            // 'L', 'Y': Reply from server.
                            // 'R', 'P': Resent last packet, either direction.

    vxt_word length;        // Total number of bytes in this block, incl header.
    vxt_word notlength;     // ~Length (used for checking).
    vxt_word cmd;           // Command to execute / result.
    vxt_word _;             // Machine ID of block sender. NOT USED!
    vxt_word __;            // Machine ID of intended reciever. NOT USED!
    vxt_word process_id;    // ID of sending process. (for closeall)
    vxt_dword crc32;        // CRC-32 of entire block. (calculated with crc32 set to zero)
    vxt_byte data[];        // Payload data.
});

enum {
    IFS_RMDIR,
    IFS_MKDIR,
    IFS_CHDIR,
    IFS_CLOSEFILE,
    IFS_COMMITFILE,
    IFS_READFILE,
    IFS_WRITEFILE,
    IFS_LOCKFILE,
    IFS_UNLOCKFILE,
    IFS_GETSPACE,
    IFS_SETATTR,
    IFS_GETATTR,
    IFS_RENAMEFILE,
    IFS_DELETEFILE,
    IFS_OPENFILE,
    IFS_CREATEFILE,
    IFS_FINDFIRST,
    IFS_FINDNEXT,
    IFS_CLOSEALL,
    IFS_EXTOPEN,
    IFS_ENDOFLIST
};

VXT_PIREPHERAL(rifs, {
    int irq;
    vxt_word base_port;
    vxt_byte registers[8];

    vxt_byte buffer_input[BUFFER_SIZE];
    int buffer_input_len;

    vxt_byte buffer_output[BUFFER_SIZE];
    int buffer_output_len;
})
/*
// Expects the 'length', 'cmd' and 'process_id' to be set by caller.
static bool server_response(struct rifs *fs, struct rifs_packet *data) {
    if (fs->buffer_input_len)
        return false;

    assert(data->length <= BUFFER_SIZE);
    fs->buffer_input_len = data->length;

    struct rifs_packet *dest = (struct rifs_packet*)fs->buffer_input;
    memmov(dest, data, data->length);
    dest->notlength = ~dest->length;

    dest->packetID[0] = 'K'; dest->packetID[1] = 'Y';

    dest->crc32 = 0;
    dest->crc32 = crc32(fs->buffer_input, dest->length);

    return true;
}
*/

static bool verify_packet(struct rifs_packet *pk) {
    vxt_word len = ~pk->notlength;
    if (pk->length != len) {
        fprintf(stderr, "RIFS packet of invalid size!\n");
        return false;
    }

    vxt_dword crc = pk->crc32;
    pk->crc32 = 0;

    if (crc != crc32((vxt_byte*)pk, len)) {
        fprintf(stderr, "RIFS packet CRC failed!\n");
        return false;
    }

    pk->crc32 = crc; // Restore the CRC if needed later.
    return true;
}

static void process_packet(struct rifs *fs, struct rifs_packet *pk) {
    (void)fs;
    int data_size = pk->length - sizeof(struct rifs_packet);

    switch (pk->cmd) {
        case IFS_FINDFIRST:
            printf("PATH: %s\n", pk->data+2);
            break;
        default:
            fprintf(stderr, "Unknown RIFS command: 0x%X (payload size %d)\n", pk->cmd, data_size);
    }
}

static vxt_byte pop_data(struct vxt_pirepheral *p) {
    VXT_DEC_DEVICE(fs, rifs, p);
    assert(fs->buffer_input_len > 0);
    vxt_byte data = *fs->buffer_input;
    memmove(fs->buffer_input, &fs->buffer_input[1], --fs->buffer_input_len);
    return data;
}

static vxt_byte in(struct vxt_pirepheral *p, vxt_word port) {
    VXT_DEC_DEVICE(fs, rifs, p);
	vxt_word reg = port & 7;
	switch (reg) {
        case 0: // Serial Data Register
            return fs->buffer_input_len ? pop_data(p) : 0;
        case 5: // Line Status Register
        {
            vxt_byte res = 0x60; // Always assume transmition buffer is empty.
            if (fs->buffer_input_len)
                res |= 1;
            return res;
        }
	}
	return fs->registers[reg];
}

static void out(struct vxt_pirepheral *p, vxt_word port, vxt_byte data) {
    VXT_DEC_DEVICE(fs, rifs, p);
	vxt_word reg = port & 7;
	fs->registers[reg] = data;

	switch (reg) {
        case 0: // Serial Data Register
            if (fs->buffer_output_len >= BUFFER_SIZE) {
                fprintf(stderr, "Invalid RIFS buffer state!\n");
                fs->buffer_output_len = 0;
                return;
            }

            fs->buffer_output[fs->buffer_output_len++] = data;
            if (fs->buffer_output_len >= (int)sizeof(struct rifs_packet)) {
                struct rifs_packet *p = (struct rifs_packet*)fs->buffer_output;
                if (p->length == fs->buffer_output_len) { // Is packet ready?
                    fs->buffer_output_len = 0;
                    if (verify_packet(p))
                        process_packet(fs, p);
                }
            }
            break;
        case 3: // Line Control Register
            fprintf(stderr, "LCR write! Assume RIFS state reset.\n");
            fs->buffer_output_len = fs->buffer_input_len = 0;
            break;
    }
}

static vxt_error install(vxt_system *s, struct vxt_pirepheral *p) {
    VXT_DEC_DEVICE(fs, rifs, p);
    vxt_system_install_io(s, p, fs->base_port, fs->base_port + 7);
    return VXT_NO_ERROR;
}

static vxt_error destroy(struct vxt_pirepheral *p) {
    vxt_system_allocator(VXT_GET_SYSTEM(rifs, p))(p, 0);
    return VXT_NO_ERROR;
}

static vxt_error reset(struct vxt_pirepheral *p) {
    VXT_DEC_DEVICE(fs, rifs, p);
    vxt_memclear(fs->registers, sizeof(fs->registers));
    return VXT_NO_ERROR;
}

static const char *name(struct vxt_pirepheral *p) {
    (void)p;
    return "RIFS Server";
}

struct vxt_pirepheral *vxtp_rifs_create(vxt_allocator *alloc, vxt_word base_port, int irq) {
    struct vxt_pirepheral *p = (struct vxt_pirepheral*)alloc(NULL, VXT_PIREPHERAL_SIZE(rifs));
    vxt_memclear(p, VXT_PIREPHERAL_SIZE(rifs));
    VXT_DEC_DEVICE(fs, rifs, p);

    fs->base_port = base_port;
    fs->irq = irq;

    p->install = &install;
    p->destroy = &destroy;
    p->name = &name;
    p->reset = &reset;
    p->io.in = &in;
    p->io.out = &out;
    return p;
}

File: modules/lua/rifs.lua
-- Copyright (c) 2019-2024 Andreas T Jonsson <mail@andreasjonsson.se>
--
-- This software is provided 'as-is', without any express or implied
-- warranty. In no event will the authors be held liable for any damages
-- arising from the use of this software.
--
-- Permission is granted to anyone to use this software for any purpose,
-- including commercial applications, and to alter it and redistribute it
-- freely, subject to the following restrictions:
--
-- 1. The origin of this software must not be misrepresented; you must not
--    claim that you wrote the original software. If you use this software
--    in a product, an acknowledgment in the product documentation would be
--    appreciated but is not required.
--
-- 2. Altered source versions must be plainly marked as such, and must not be
--    misrepresented as being the original software.
--
-- 3. This notice may not be removed or altered from any source
--    distribution.

local base_port = tonumber(({...})[1])
local registers = {0, 0, 0, 0, 0, 0, 0}
local server = {
    input_buffer = {},
    output_buffer = {}
}

function set_error(fmt, ...)
    local e = string.format(fmt, ...)
    log(e)
    server.error = e
end

function log(fmt, ...)
    print(string.format("[%s] " .. fmt, name(), ...))
end

function name()
    return "RIFS Server"
end

function io_in(port)
    if server.error then
        -- We are in a fatal error state. Wait for reset.
        return
    end

    local reg = port - base_port
    if reg == 0 then -- Serial Data Register
        if server.dlab then
            return registers[reg]
        end
        return table.remove(server.input_buffer) or 0
    elseif reg == 5 then -- Line Status Register
        local res = 0x60 -- Always assume transmition buffer is empty.
        if #server.input_buffer > 0 then
            res = res | 1
        end
        return res
    end
    return 0
end

function io_out(port, data)
    local tm = os.clock()
    local reg = port - base_port

    if reg == 0 and not server.error then -- Serial Data Register
        if server.dlab then
            return
        end

        table.insert(server.output_buffer, data)

        local sz = #server.output_buffer
        if sz >= 18 then
            if server.current_packet == nil then
                local str = string.pack(string.rep("B", sz), table.unpack(server.output_buffer))
                server.current_packet = {
                    id = string.sub(str, 1, 2),
                    length = string.unpack("H", str, 3),
                    notlength = string.unpack("H", str, 5),
                    cmd = string.unpack("H", str, 7),
                    process_id = string.unpack("H", str, 13),
                    crc32 = string.unpack("I4", str, 15)
                }
            end

            if server.current_packet.length == sz then
                if verify_packet(server.current_packet, server.output_buffer) then
                    local payload = string.pack(string.rep("B", sz - 18), select(19, table.unpack(server.output_buffer)))
                    process_packet(server.current_packet, payload)
                end
                server.output_buffer = {}
            end
        end
    elseif reg == 3 then -- Line Control Register
        local dlab = (data & 0x80) ~= 0
        if server.dlab and not dlab then
            server.error = nil
            server.input_buffer = {}
            server.output_buffer = {}
            log("Reset!")
        end
        server.dlab = dlab
    end

    wait((os.clock() - tm) * frequency())
end

function install()
    if not base_port then
        base_port = 0x178
        log("Default IO port: 0x%X", base_port)
    end
    install_io(base_port, base_port + 7)
end

function verify_packet(header, data)
    if header.length ~= (~header.notlength) & 0xFFFF then
        set_error("Invalid packet length!")
        return false
    end

    -- Reset CRC32
    data[15] = 0
    data[16] = 0
    data[17] = 0
    data[18] = 0

    local str = string.pack(string.rep("B", #data), table.unpack(data))
    if header.crc32 ~= math.crc32(str) then
        set_error("Invalid CRC!")
        return false
    end
    return true
end

function clone_header(header)
    local p = {}
    for k,v in pairs(header) do
        p[k] = v
    end
    return p
end

function process_packet(header, payload)
    local func = rifs_commands[header.cmd]
    if not func then
        log("Unknown RIFS command: 0x%X (payload size %d)", header.cmd, #payload)

        header = clone_header(header)
        header.cmd = 0x16 -- Unknown command
        server_response(header)
        return
    end

    local cmd = func(header, payload)
    if server.error then
        return
    end

    if type(cmd) == "number" then
        header = clone_header(header)
        header.cmd = 0
        server_response(header)
    end
end

function server_response(header, payload)
    if #server.input_buffer ~= 0 then
        set_error("Expected input buffer to be empty!")
        return
    end

    local payload_size = payload and #payload or 0

    header.id = "LY"
    header.length = 18 + payload_size
    header.notlength = (~header.length) & 0xFFFF

    local data = string.format("%s%s%s%s%s%s%s%s",
        header.id,
        string.pack("H", header.length),
        string.pack("H", header.notlength),
        string.pack("H", header.cmd),
        string.pack("I4", 0), -- Unused
        string.pack("H", header.process_id),
        string.pack("I4", 0), -- CRC32
        payload
    )

    server.input_buffer = {string.unpack(string.rep("B", #data), data)}

    local crc = string.pack("I4", math.crc32(data))
    server.input_buffer[15] = crc[1]
    server.input_buffer[16] = crc[2]
    server.input_buffer[17] = crc[3]
    server.input_buffer[18] = crc[4]
end

--------------------------------------------------------------------------------
-- Implementation of RIFS commands
--------------------------------------------------------------------------------

function rifs_rmdir(header, payload)
    return 0 -- Just return OK
end

rifs_commands = {
    [0x0] = rifs_rmdir
}
